src/olmo_core/train/callbacks/batch_size_scheduler.py

```diff
@@ -0,0 +1,140 @@
+
+        for next_batch_size, next_event_start in self.schedule[1:]:
+            if next_event_start.value <= event_start.value:
+                raise OLMoConfigurationError(
+                    "subsequent events in the batch size schedule must be configured to occur after previous events"
+                )
+
+            if next_batch_size > batch_size:
+                if next_batch_size % batch_size != 0:
+                    raise OLMoConfigurationError(
+                        "invalid batch size schedule, batch size can only be increased by an integer factor"
+                    )
+            elif next_batch_size < batch_size:
+                if batch_size % next_batch_size != 0:
+                    raise OLMoConfigurationError(
+                        "invalid batch size schedule, batch size can only be decreased by an integer factor"
+                    )
+            else:
+                raise OLMoConfigurationError(
+                    "invalid batch size schedule, duplicate batch size in schedule"
+                )
+
+            batch_size = next_batch_size
+            event_start = next_event_start
+
+    @property
+    def current_batch_size(self) -> int:
+        return self.trainer.data_loader.global_batch_size
+
+    def post_attach(self):
+        self._maybe_update_batch_size_and_lr()
+
+    def post_checkpoint_loaded(self):
+        self._maybe_update_batch_size_and_lr()
+
+    def pre_load_batch(self):
+        self._maybe_update_batch_size_and_lr()
+
+    def _maybe_update_batch_size_and_lr(self):
+        # Find latest event in the schedule to apply.
+        for target_batch_size, event_start in reversed(self.schedule):
+            if event_start.due(
+                step=self.trainer.global_step,
+                tokens=self.trainer.global_train_tokens_seen,
+                epoch=self.trainer.epoch,
+            ):
+                if target_batch_size > self.current_batch_size:
+                    assert target_batch_size % self.current_batch_size == 0
+                    # When increasing the batch size, need to wait until 'batches_processed' is divisible
+                    # by the factor that batch size is being increased by.
+                    ratio = target_batch_size // self.current_batch_size
+                    if self.trainer.data_loader.batches_processed % ratio != 0:
+                        self._update_batch_size_and_lr(target_batch_size)
+                elif target_batch_size < self.current_batch_size:
+                    assert self.current_batch_size % target_batch_size == 0
+                    self._update_batch_size_and_lr(target_batch_size)
+
+                break
+
+    def _update_batch_size_and_lr(self, batch_size: int):
+        log.info(f"Changing global batch size to {batch_size:,d}...")
+        ratio = batch_size / self.current_batch_size
+        lr_adjustment_factor = math.sqrt(ratio)
```

Comment:
It's worth noting that this is true for Adam, but may not be true for other optimizers. I don't know how to package that fact in the code. Can we check somehow whether we have Adam, and throw an error if we don't?

Reply:
29a9ae02